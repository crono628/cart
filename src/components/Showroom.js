import React from 'react';
import Item from './Item';
import uniqid from 'uniqid';

const Showroom = ({ onAdd, inventory }) => {
  const products = inventory.map((product) => {
    return <Item key={uniqid()} product={product} onAdd={onAdd} />;
  });

  return (
    <div className="showroom-title">
      <h1>Products</h1>
      <div className="showroom">{products}</div>
    </div>
  );
};

export default Showroom;
