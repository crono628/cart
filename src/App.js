import React, { useState } from 'react';
import { Routes, Route } from 'react-router-dom';
import data from './components/data';
import Cart from './components/Cart';
import Nav from './components/Nav';
import Showroom from './components/Showroom';
import Welcome from './components/Welcome';

const App = () => {
  const [cart, setCart] = useState([]);

  const onAdd = (product) => {
    const inBag = cart.find((item) => item.id === product.id);
    let copy = [...cart];
    if (inBag) {
      setCart(
        copy.map((item) =>
          item.id === product.id ? { ...inBag, qty: inBag.qty + 1 } : item
        )
      );
    } else {
      setCart([...copy, { ...product, qty: 1 }]);
    }
  };

  const onRemove = (product) => {
    const inBag = cart.find((item) => item.id === product.id);
    let copy = [...cart];
    if (inBag.qty === 1) {
      setCart(copy.filter((items) => items.id !== product.id));
    } else {
      setCart(
        copy.map((item) =>
          item.id === product.id ? { ...inBag, qty: inBag.qty - 1 } : item
        )
      );
    }
  };

  const cartQuantity = cart.reduce((a, b) => a + b.qty, 0);

  return (
    <div>
      <Nav quantity={cartQuantity} />
      <div className="container">
        <Routes>
          <Route path="/" element={<Welcome />} />
          <Route
            path="/showroom"
            element={<Showroom inventory={data} onAdd={onAdd} />}
          />
          <Route
            path="/shopping-cart"
            element={<Cart onRemove={onRemove} onAdd={onAdd} items={cart} />}
          />
        </Routes>
      </div>
    </div>
  );
};

export default App;
