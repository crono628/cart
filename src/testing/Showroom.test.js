import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Showroom from '../components/Showroom';
import App from '../App';

const data = [
  {
    name: 'Sony Betamax',
    src: 'https://www.extremetech.com/wp-content/uploads/2015/11/772066351852676897.jpg',
    description: 'The future of home video, forever',
    price: 999,
    id: 1,
  },
];

afterEach(cleanup);

describe('Showroom', () => {
  it('data renders and has a buy button', () => {
    const onAdd = jest.fn();
    const { getByTestId, container, debug } = render(
      <Showroom onAdd={onAdd} inventory={data} />
    );
    const buyButton = getByTestId('buy-button');
    expect(buyButton.tagName).toBe('BUTTON');
    expect(buyButton.textContent).toBe('Add To Cart');
    fireEvent.click(buyButton);
    expect(onAdd).toBeCalledTimes(1);
    fireEvent.click(buyButton);
    expect(onAdd).toBeCalledTimes(2);
    expect(container.firstChild).toMatchSnapshot();
  });
});
