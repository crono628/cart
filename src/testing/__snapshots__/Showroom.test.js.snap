// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Showroom data renders and has a buy button 1`] = `
<div
  class="showroom-title"
>
  <h1>
    Products
  </h1>
  <div
    class="showroom"
  >
    <div
      class="showroom-item"
    >
      <img
        src="https://www.extremetech.com/wp-content/uploads/2015/11/772066351852676897.jpg"
      />
      <h4>
        Sony Betamax
      </h4>
      <h6>
        The future of home video, forever
      </h6>
      <div>
        $
        999
      </div>
      <button
        data-testid="buy-button"
      >
        Add To Cart
      </button>
    </div>
  </div>
</div>
`;
